core/drag/drag-manager.js

```diff
@@ -0,0 +1,1314 @@
+                }
+
+                draggingOperationContext.draggedImage = this._sanitizeDraggedImage(
+                    draggedImage
+                );
+
+                this._addTranslateListeners();
+                draggingOperationContext.isDragging = true;
+                this._rootComponent.needsDraw = true;
+            } else {
+                this._translateComposer._cancel();
+            }
+        }
+    },
+
+    handleTranslate: {
+        value: function (event) {
+            this._draggingOperationContext.deltaX = event.translateX;
+            this._draggingOperationContext.deltaY = event.translateY;
+            this._draggingOperationContext.positionX = (
+                this._draggingOperationContext.startPositionX + event.translateX
+            );
+            this._draggingOperationContext.positionY = (
+                this._draggingOperationContext.startPositionY + event.translateY
+            );
+            this._rootComponent.needsDraw = true;
+        }
+    },
+
+    handleTranslateEnd: {
+        value: function () {
+            this._willTerminateDraggingOperation = true;
+            this._rootComponent.needsDraw = true;
+        }
+    },
+
+    handleTranslateCancel: {
+        value: function () {
+            this.draggingOperationContext.currentDropTarget = null;
+            this._willTerminateDraggingOperation = true;
+            this._rootComponent.needsDraw = true;
+        }
+    },
+
+    /**
+     * Draw Cycles Management
+     */
+
+    willDraw: { 
+        value: function () {
+            var draggingOperationContext;
+
+            if (
+                (draggingOperationContext = this._draggingOperationContext) &&
+                draggingOperationContext.isDragging
+            ) {
+                var draggable = draggingOperationContext.draggable;
+
+                if ( 
+                    !this._draggedImageBoundingRect && 
+                    draggingOperationContext.draggable
+                ) {
+                    this._draggedImageBoundingRect = (
+                        draggingOperationContext.draggable.element.getBoundingClientRect()
+                    );
+    
+                    if (draggable.draggableContainer) {
+                        this._draggableContainerBoundingRect = (
+                            draggable.draggableContainer.getBoundingClientRect()
+                        );
+                    }
+                } else {
+                    var droppable = this._findDroppableAtPosition(
+                        draggingOperationContext.positionX,
+                        draggingOperationContext.positionY
+                    );
+
+                    if (droppable && 
+                        !draggingOperationContext.dropTargetCandidates.has(
+                            droppable
+                        )
+                    ) {
+                        droppable.classList.add('invalid-drop-target');
```

Comment:
The DOM is not supposed to be touched in willDraw. Do we anticipate this logic to potentially trigger some components to be added to the current draw loop? If not, it might be better to move it to the beginning of draw()

Reply:
The DOM is not touched here. the droppable variable is a pointer to a component so it will be automatically added/removed here:
https://github.com/montagejs/montage/blob/master/ui/component.js#L3329